android/java/org/chromium/chrome/browser/crypto_wallet/util/WalletUtils.java

```diff
@@ -17,42 +19,60 @@
 import org.chromium.brave_wallet.mojom.TxDataUnion;
 import org.chromium.chrome.R;
 
+import java.nio.ByteBuffer;
 import java.util.ArrayList;
 import java.util.Arrays;
+import java.util.HashSet;
 import java.util.List;
+import java.util.Set;
+import java.util.stream.Collectors;
 
 public class WalletUtils {
-    public static String getUniqueNextAccountName(Context context, AccountInfo[] accountInfos,
-            String cryptoAccountType, @CoinType.EnumType int coin) {
-        List<AccountInfo> accountInfoList = new ArrayList<>();
-        for (AccountInfo accountInfo : accountInfos) {
-            if (accountInfo.coin == coin) {
-                accountInfoList.add(accountInfo);
-            }
+    private static String getNewAccountPrefixForCoin(@CoinType.EnumType int coinType) {
+        switch (coinType) {
+            case CoinType.ETH:
+                return "Ethereum";
+            case CoinType.SOL:
+                return "Solana";
+            case CoinType.FIL:
+                return "Filecoin";
+            case CoinType.BTC:
+                return "Bitcoin";
         }
-        return getUniqueNextAccountName(context, accountInfoList.toArray(new AccountInfo[0]),
-                1 /* account name 1..n*/, cryptoAccountType);
+        assert false;
+        return "";
     }
 
-    private static String getUniqueNextAccountName(
-            Context context, AccountInfo[] accountInfos, int number, String cryptoAccountTypeInfo) {
-        String accountName = context.getString(R.string.new_account_prefix, cryptoAccountTypeInfo,
-                String.valueOf(accountInfos.length + number));
-        for (AccountInfo accountInfo : accountInfos) {
-            if (accountInfo.name.equals(accountName)) {
-                return getUniqueNextAccountName(
-                        context, accountInfos, number + 1, cryptoAccountTypeInfo);
+    public static String generateUniqueAccountName(
+            Context context, @CoinType.EnumType int coinType, AccountInfo[] accountInfos) {
+        Set<String> allNames = Arrays.stream(accountInfos)
+                                       .map(acc -> acc.name)
+                                       .collect(Collectors.toCollection(HashSet::new));
+
+        for (int number = 1; number < 1000; ++number) {
+            String accountName = context.getString(R.string.new_account_prefix,
+                    getNewAccountPrefixForCoin(coinType), String.valueOf(number));
+
+            if (!allNames.contains(accountName)) {
+                return accountName;
             }
         }
-        return accountName;
+        return "";
+    }
+
+    public static boolean accountIdsEqual(AccountId left, AccountId right) {
+        return left.serialize().equals(right.serialize());
```

Comment:
I would recommend to compare actual values/fields in `AccountId` than comparing `Message` object, for clarity and to avoid any unexpected issue/edge-case, not sure how accurate mojo `Message` object equality is on different devices or variants.

Reply:
R&D for later: maybe mojo could generate an equals method like [bool Equals(const Employee& other);](https://www.cnblogs.com/huangguanyuan/p/10320950.html)

Reply:
The idea not to use comparison by fields is that we have to support it separately for mojom interface declaration which means they can easily become out of sync. So it would need a cumbersome test like the one I removed for AccountInfo.

Serialization format is pretty straightforward and should always produce equal binary arrays for equal structs:
```
    @SuppressWarnings("unchecked")
    @Override
    protected final void encode(org.chromium.mojo.bindings.Encoder encoder) {
        org.chromium.mojo.bindings.Encoder encoder0 = encoder.getEncoderAtDataOffset(DEFAULT_STRUCT_INFO);
        encoder0.encode(this.coin, 8);
        encoder0.encode(this.kind, 12);
        encoder0.encode(this.keyringId, 16, false);
        encoder0.encode(this.address, 24, false);
    }
```

Reply:
if you are 100% sure that `it should always produce equal binary arrays for equal structs` then It's okay to have this though would suggest to add a reference if available.